### DISCUSSION TITLE
The same command run on different systems (environments) possibably outputs different results

### DISCUSSION THREAD

**Discussion by fu123456** [ImageMagick/ImageMagick#4878] [General]
The command line ' convert <input> <output>' I often used recently made me doubt.
I run 'convert test.exr output.jpg' on my Manjaro (Archlinux) system, and obtain the different result with that produced on other sysytems (environments). Note that, exr image indicate high dynamic range (HDR) image. For fast understand, please see the following figure.
![位图](https://user-images.githubusercontent.com/22411577/155126163-53c39a72-c1d2-448b-9f7c-6aa869e19b06.png)
In the above figure, (a) Input exr image (visually showed by Photoshop or Gimp for easy observation), (b) my result produced on my Manjaro system, (c) the correct result produced on other environments such as Ubuntu 16 and Centos (I have tested some different system). Thus, basicually, we can argue that the result in (c) should be correct and that in (b) should be incorrect. I run the command line on my Manjaro system without any error. Running this command line by convert v7 (Imagemagick7) and convert v6 (Imagemagick6) suffers from the same issue. So I do not know why like this and this make me doubt the stability and robustness of the Imagemagick software.

**Comment by urban-warrior** [comment]
Can you post your original EXR image here along with the ImageMagick command you used?  We need to reproduce the problem before we can comment further.

**Reply by fu123456** [reply]
here https://drive.google.com/file/d/1IgSllkNopmxPkJd0QGiOKp4q_XI2krOn/view?usp=sharing

**Comment by snibgo** [comment]
@fu123456: As you say, the EXR file is high dynamic range. What are you converting it to? If you are converting to LDR (Low Dynamic Range) such as JPG, do you simply want to clip the high values? Or do you want to rescale all pixels to bring the range down to LDR?

What software and hardware are you using to view the images ? Does that software and hardware show HDR correctly? Or does it convert to LDR?

EDIT: Check that your IM has HDRI. If it doesn't, HDR values will always be clipped.

**Reply by fu123456** [reply]
The Exr image can be viewed by Photoshop and Gimp well. I am sure that 'convert *.exr *.jpg' can do this well. Some researchers use this tool to convert image format simply. Now, I also am sure that the results produced on my system and PC are incorrect, while the results tested on other systems such as Ubuntu 16 and centos 7 are correct. Note that, here 'correct' and 'incorrect' are made according to my judgment on the comparison of the results.

**Reply by fmw42** [reply]
The image you posted is a normal 16-bit range image. It is not high dynamic range as the the gray levels do not exceed the normal 16-bit range.  My results on IM 7.1.0-26 Q16 HDRI Mac OSX looks like your B) result. I have no reason to think that is wrong.  I am using OpenEXR 2.3.0 and libjpeg-turbo 2.1.0

What processing did you do to make such a small size EXR file?  Did you crop it in some other tool?

**Reply by fu123456** [reply]
A graphics rendering tool, but this is not important. In your Mac OSX system, you can use Photoshop or Gimp to open this exr file (here https://drive.google.com/file/d/1IgSllkNopmxPkJd0QGiOKp4q_XI2krOn/view?usp=sharing) and visually compare it with one produced by the command of 'convert', and see whether the two are the same? If they are not same, how to explain this.

**Reply by fmw42** [reply]
Perhaps PS is automatically stretching the data to full dynamic range.

**Reply by fmw42** [reply]
Imagemagick use OpenEXR to read the file. I do not know what software PS uses.

**Comment by fmw42** [comment]
What are the IM versions and versions of libjpeg and libexr used on the different systems are they the same versions of all software?

**Reply by fu123456** [reply]
In my Manjaro system, I have installed the following packages (libs) and their versions:
openexr-3.1.4-1
lib32-openexr-3.1.4-1
libjpeg-turbo=2.1.1

**Reply by fmw42** [reply]
Are these the same versions as used on your other systems?

**Reply by fu123456** [reply]
No. Other systems should be with lower versions. I also use Ubuntu 16 with several package updating during several years, which can consistently produce the correct (normal) results at least I feel.

**Comment by urban-warrior** [comment]
HDR images are linear light.  ImageMagick does not auto correct colors, it leaves that to the user.  Add `-gamma 2.2` to your command-line to get expected results.

**Reply by fmw42** [reply]
Thanks. That helps clarify it.

**Reply by fu123456** [reply]
> HDR images are linear light. ImageMagick does not auto correct colors, it leaves that to the user. Add `-gamma 2.2` to your command-line to get expected results.
How to explain the different results on different system? If I am not sure to use '-gamma 2.2' on different systems, how to decide to whether use this parameter?

**Comment by snibgo** [comment]
fmw42 wrote:
> The image you posted is a normal 16-bit range image. It is not high dynamic range as the the gray levels do not exceed the normal 16-bit range.

According to IM v7.1.0-4, with HDRI, the blue channel does exceed the normal range. I have named the input file "blueCar.exr". Commands are Windows BAT syntax.

```
f:\web\im>C:\IM\ImageMagick-7.1.0-4-Q16-HDRI\magick blueCar.exr -verbose info:
Image:
  Filename: blueCar.exr
  Format: EXR (High Dynamic-range (HDR))
  Class: DirectClass
  Geometry: 256x256+0+0
  Units: Undefined
  Colorspace: RGB
  Type: TrueColorAlpha
  Base type: Undefined
  Endianness: Undefined
  Depth: 16-bit
  Channel depth:
    Red: 16-bit
    Green: 16-bit
    Blue: 16-bit
    Alpha: 1-bit
  Channel statistics:
    Pixels: 65536
    Red:
      min: 1134.98  (0.0173187)
      max: 48799.3 (0.744629)
      mean: 11448.7 (0.174697)
      median: 7031.89 (0.1073)
      standard deviation: 6587.29 (0.100516)
      kurtosis: 2.45064
      skewness: 1.71678
      entropy: 0.888489
    Green:
      min: 1303.98  (0.0198975)
      max: 58399.1 (0.891113)
      mean: 15249.4 (0.23269)
      median: 9615.85 (0.146729)
      standard deviation: 8876.66 (0.135449)
      kurtosis: 2.08806
      skewness: 1.73126
      entropy: 0.881749
    Blue:
      min: 1459.98  (0.0222778)
      max: 98046.5 (1.49609)
      mean: 25115.4 (0.383237)
      median: 14247.8 (0.217407)
      standard deviation: 18478.4 (0.281963)
      kurtosis: 1.35533
      skewness: 1.66064
      entropy: 0.848896
```
The maximum blue value is nearly 50% above the norm. If we clip ("-clamp") those values, those pixels will change in hue.

What proportion of pixels are outside the norm?
```
magick ^
  blueCar.exr ^
  ( +clone -clamp ) ^
  -compose Difference -composite ^
  -threshold 0 ^
  -format "%%[fx:mean]\n" info:

0.0517426
```
About 5% of pixels would be clipped, and thus change in hue.

exiftool reports that blueCar.exr has sRGB primaries. As urban-warrior says, exr files are assumed to be linear.

If we simply "magick blueCar.jpg x.jpg" IM will automatically convert the file from linear to sRGB (roughly with a power curve).

![x](https://user-images.githubusercontent.com/33812368/155268376-c472593e-0674-4123-83b0-50ded330759a.jpg)

If we instead convert to png format, IM won't convert to sRGB, but will set the file metadata gamma to 1. Some software (such as Windows Photo Viewer) will correctly convert from linear to sRGB. A a general rule, I wouldn't assume that all viewers do this correctly, and I suggest converting explicitly with "-set colorspace RGB -colorspace sRGB".

If we want to avoid clipping and thus changes to hue, we can divide all values by the maximum, so the largest value is then 100% of normal, without changing the ratios between the channels.
```
magick ^
  blueCar.exr ^
  -channel RGB ^
  -evaluate Divide %%[fx:maxima] ^
  +channel ^
  -set colorspace RGB -colorspace sRGB ^
  -verbose +write info: ^
  x2.jpg
```
![x2](https://user-images.githubusercontent.com/33812368/155268318-5849cffa-5bc2-4dbd-827e-770c7d4ed620.jpg)
x2.jpg is slightly darker than x.jpg.

**Reply by fmw42** [reply]
My mistake. I scanned the verbose info too quickly.

**Comment by snibgo** [comment]
> How to explain the different results on different system?

@fu123456: please show the exact commands you used, and the output image files. Not screenshots, but the actual files.

Also show the text output of "convert -version" from each version of ImageMagick.

**Reply by fu123456** [reply]
The command lines are the same. It is like 'convert input.exr ouput.jpg'. On my Manjaro system, I use convert v7 and convert v6, and obtain the same output. On other systems, I use convert v6.  The images are available at https://drive.google.com/file/d/1UjYjbGUDFsQXbRMpcn_yHPhSmbAk5EyQ/view?usp=sharing
